Report minimum storage threshold targets

dotnwat opened the pull request:
Disk space management policy needs to be able to juggle space between the disk cache and log storage. To do this some thresholds will be useful, such as a minimum amount of space that log storage needs for normal functionality, and the minimum amount it needs to meet some basic policies like data retention goals.

This PR adds reporting for
* minimum needed capacity
* minimum capacity needed for `retention.bytes` policies
* minimum capacity needed for `retention.ms` policies

(note that I realize `retention.{bytes,ms}` refers to something different than local retention. in this pr we use the normalized version which for cloud storage will be local retention. i believe i've been using this terminology rather loosely).

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.1.x
- [ ] v22.3.x
- [ ] v22.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* none

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1970,26 +1970,112 @@ disk_log_impl::disk_usage_target(gc_config cfg, usage usage) {
+      [](storage::usage acc, storage::usage u) { return acc + u; });
+
+    // extrapolate out for the missing period of time in the retention period
+    auto missing_bytes = (usage.total() * missing.value()) / duration.value();
```
I think this is saying "I expect to see same throughput that ingested the latest segments for the entire remainder of the retention period". This seems like a reasonable, but it makes me wonder about bursty workloads, where a large number of records are written in the span of a few dozens of seconds, and then stopped. In such a case, we could end up with a small `duration` and large `missing` and end up with a very large value for `missing_bytes`.

Maybe it makes sense to compute the expected throughput with `now() - start_timestamp` instead of `end_timestamp - start_timestamp`, though given these are user-input timestamps, it's probably best to avoid using `now()`

I suppose regardless it might be better to overestimate the wanted bytes anyway for the goal of avoiding out of space issues

dotnwat replied:
I attempted to capture this issue with a heuristic like not making any estimate unless we have say 10 seconds worth of data. But this is going to just be a whack-a-mole game I'm afraid. Perhaps we would do well to do add a cap here, too, such as double or triple the amount of data currently on disk. This may work well since the goal isn't necessarily to observe the system once and know how much space we need. Rather, this will be monitoring periodically. So capping will let the estimate continue to portray growth requirement, while also eliminating huge bursts?

abhijat commented on src/v/storage/types.h:
```diff
@@ -467,19 +467,71 @@ struct usage {
+    size_t min_capacity{0};
+    size_t min_capacity_wanted{0};
+
+    friend usage_target operator+(usage_target lhs, const usage_target& rhs) {
```
nit: is `friend` required here, or can this just be a simple operator overload, as both fields in `usage_target` are public?

dotnwat replied:
@abhijat you are right friend is not needed. It's a habit to do that in the tree because I believe there is some sort of compilation time benefit. Never really dug into the details.

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1807,10 +1807,13 @@ log make_disk_backed_log(
+ * assumes that the compaction gate is held.
+ */
+ss::future<std::pair<usage, reclaim_size_limits>>
+disk_log_impl::disk_usage_and_reclaim(gc_config cfg) {
```
fixed

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1908,6 +1908,19 @@ disk_log_impl::disk_usage_and_reclaim(gc_config cfg) {
     co_return std::make_pair(usage, reclaim);
 }
 
+/*
+ * assumes that the compaction gate is held.
+ */
```
Oh that'd be cool. Unfortunately I don't think there is: short of passing around parameters, we don't have fiber-local storage to track that sort of thing. the comment is probably sufficient here since these methods are all private to the implementation.

dotnwat commented on tests/rptest/tests/full_disk_test.py:
```diff
@@ -335,6 +335,35 @@ def check(self, threshold):
+            reported = admin.get_local_storage_usage(
+                node)["target_min_capacity"]
+
+            # 1 partition with 1gb segment size
+            # 4 partition with default segment size
+            # controller partition has default segment size
+            # reservation will be for min_segments
+            expected = min_segments * ((1 * 2**30) + \
+                       (4 * default_segment_size) + \
+                       (1 * default_segment_size))
```
I think this is already happening: the `2**30` is the custom segment size for one of the automatically created topics. I've replaced this integer with `self.topics[0].segment_bytes` to make that clearer.

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1919,13 +1919,74 @@ disk_log_impl::disk_usage_and_reclaim(gc_config cfg) {
+         * if there is no delete policy enabled for a compacted topic then its
+         * local capacity requirement is limited only by compaction process and
+         * how much data is written. for this case we use a heuristic of to
+         * report space wanted as `factor * current capacity` to reflect that we
```
>The first two scenarios are covered by the min_capacity lower bound. I'm not sure how realistic the latter is, but curious if you'd thought about e.g. an upper bound on the wanted capacity or something similar

Yes, @jcsp and I spoke briefly about this and concluded that an upper bound may make sense (at least the capacity of the disk) but that reporting a huge value is also an accurate description of what "nice to have" is trying to portray. The take away is that we're gonna have to take this scenario into account at some point, but if that is here at a low level or higher up the policy food chain remains an open question.

dotnwat commented:
Force-pushed to resolve merge conflicts in admin_server.

dotnwat commented:
Force-push:

1. Added some extra trace logging
2. Updated the test to be more stable. Before I was using some fixed produces with sleeps to try to achieve some predictable throughput. But this was very noisy. Now this requests the kafka producer tool to write at a specific rate.

dotnwat commented:
Force-push: fix python formatting

dotnwat commented:
ping @andrwng when you get a chance

dotnwat commented:
Force-push: increased fuzz threshold on release it seems we write a bit faster than on debug which I was using for the calibration.

dotnwat commented:
Force-push: fix conflicts with clang-format16 changes

dotnwat commented:
Failures:
* https://github.com/redpanda-data/redpanda/issues/10502
* https://github.com/redpanda-data/redpanda/issues/10604
* https://github.com/redpanda-data/redpanda/issues/10497

andrwng reviewed (approved):
Thanks for the ping, and for clarifying. I think this looks like a good starting point as far as heuristics go. Remaining comments/clarifications aren't blocking.
